File: 4get-rs/src/scraper/unsplash.rs
use crate::config::Config;
use crate::errors::AppError;
use crate::scraper::client::HttpClient;
use crate::scraper::Scraper;
use crate::types::*;
use async_trait::async_trait;
use std::collections::HashMap;

pub struct Unsplash {
    http: HttpClient,
    api_key: String,
}

impl Unsplash {
    pub fn new(http: HttpClient, config: &Config) -> Self {
        let api_key = config
            .scrapers
            .unsplash_api_key
            .clone()
            .unwrap_or_default();
        Unsplash { http, api_key }
    }
}

#[async_trait]
impl Scraper for Unsplash {
    fn name(&self) -> &str {
        "unsplash"
    }

    async fn image(&self, query: &SearchQuery) -> Result<ImageResponse, AppError> {
        let mut filter: HashMap<String, String> = if let Some(npt) = &query.npt {
            serde_json::from_str(npt).unwrap_or_default()
        } else {
            let mut f = HashMap::new();
            f.insert("page".to_string(), "1".to_string());
            f.insert("per_page".to_string(), "20".to_string());
            f.insert("query".to_string(), query.q.clone());

            if let Some(order_by) = query.filters.get("f_order_by") {
                if order_by != "relevance" {
                    f.insert("order_by".to_string(), order_by.clone());
                }
            }
            if let Some(orientation) = query.filters.get("f_orientation") {
                if orientation != "any" {
                    f.insert("orientation".to_string(), orientation.clone());
                }
            }
            if let Some(license) = query.filters.get("f_license") {
                if license != "any" {
                    f.insert("plus".to_string(), license.clone());
                }
            }

            f
        };

        let search_term = filter.get("query").cloned().unwrap_or_default();
        let referer = format!(
            "https://unsplash.com/s/photos/{}",
            search_term.replace(' ', "-")
        );

        let mut req = self
            .http
            .client
            .get("https://unsplash.com/napi/search/photos")
            .query(&filter)
            .header("Accept", "*/*")
            .header("Accept-Language", "en-US")
            .header("Referer", &referer)
            .header("client-geo-region", "global")
            .header("x-client-version", "8999df28be3f138bf2c646df5d656e4dc6970ba0")
            .header("DNT", "1")
            .header("Sec-GPC", "1")
            .header("Connection", "keep-alive")
            .header("Sec-Fetch-Dest", "empty")
            .header("Sec-Fetch-Mode", "cors")
            .header("Sec-Fetch-Site", "same-origin");

        if !self.api_key.is_empty() {
            req = req.header("Authorization", format!("Client-ID {}", self.api_key));
        }

        let json = req
            .send()
            .await?
            .json::<serde_json::Value>()
            .await?;

        let mut response = ImageResponse::empty();

        let results = match json.get("results").and_then(|v| v.as_array()) {
            Some(r) => r,
            None => return Ok(response),
        };

        for image in results {
            let raw_url = image
                .get("urls")
                .and_then(|u| u.get("raw"))
                .and_then(|v| v.as_str())
                .unwrap_or("");

            if raw_url.is_empty() {
                continue;
            }

            let base = raw_url.split('?').next().unwrap_or(raw_url);

            let is_premium = image.get("premium").and_then(|v| v.as_bool()).unwrap_or(false);
            let is_plus = image.get("plus").and_then(|v| v.as_bool()).unwrap_or(false);

            let width = image.get("width").and_then(|v| v.as_u64()).unwrap_or(0) as u32;
            let height = image.get("height").and_then(|v| v.as_u64()).unwrap_or(0) as u32;

            let mut sources = Vec::new();

            if is_premium || is_plus {
                let x900_w = 900u32;
                let x900_h = ((height as f64) * (900.0 / (width as f64))) as u32;
                let x500_w = 500u32;
                let x500_h = ((height as f64) * (500.0 / (width as f64))) as u32;

                sources.push(ImageSource {
                    url: base.to_string(),
                    width: Some(width),
                    height: Some(height),
                });
                sources.push(ImageSource {
                    url: format!("{}?w=900", base),
                    width: Some(x900_w),
                    height: Some(x900_h),
                });
                sources.push(ImageSource {
                    url: format!("{}?w=500", base),
                    width: Some(x500_w),
                    height: Some(x500_h),
                });
            } else {
                let x500_w = 500u32;
                let x500_h = ((height as f64) * (500.0 / (width as f64))) as u32;

                sources.push(ImageSource {
                    url: base.to_string(),
                    width: Some(width),
                    height: Some(height),
                });
                sources.push(ImageSource {
                    url: format!("{}?w=500", base),
                    width: Some(x500_w),
                    height: Some(x500_h),
                });
            }

            let desc = image
                .get("description")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .trim();
            let alt_desc = image
                .get("alt_description")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .trim();

            let title = match (desc, alt_desc) {
                ("", "") => String::new(),
                ("", b) => b.to_string(),
                (a, "") => a.to_string(),
                (a, b) => format!("{}: {}", a, b),
            };

            let slug = image.get("slug").and_then(|v| v.as_str()).unwrap_or("");
            let page_url = format!("https://unsplash.com/photos/{}", slug);

            response.image.push(ImageResult {
                title,
                url: page_url,
                source: sources,
            });
        }

        let total_pages = json
            .get("total_pages")
            .and_then(|v| v.as_u64())
            .unwrap_or(0);

        let current_page = filter
            .get("page")
            .and_then(|v| v.parse::<u64>().ok())
            .unwrap_or(1);

        if current_page < total_pages {
            filter.insert("page".to_string(), (current_page + 1).to_string());
            response.npt = Some(serde_json::to_string(&filter).unwrap_or_default());
        }

        Ok(response)
    }
}

File: scraper/unsplash.php
<?php

class unsplash{
	
	public function __construct(){
		
		include "lib/fuckhtml.php";
		$this->fuckhtml = new fuckhtml();
		
		include "lib/backend.php";
		$this->backend = new backend("unsplash");
	}
	
	public function getfilters($page){
		
		return [
			"order_by" => [
				"display" => "Order by",
				"option" => [
					"relevance" => "Relevance",
					"latest" => "Newest",
					"editorial" => "Curated"
				]
			],
			"orientation" => [
				"display" => "Order by",
				"option" => [
					"any" => "Any orientation",
					"landscape" => "Landscape",
					"portrait" => "Portrait",
					"squarish" => "Square"
				]
			],
			"license" => [
				"display" => "License",
				"option" => [
					"any" => "Any license",
					"only" => "Unsplash+",
					"none" => "Free"
				]
			]
		];
	}
	
	private function get($proxy, $url, $get = [], $referer){
		
		$curlproc = curl_init();
		
		if($get !== []){
			$get = http_build_query($get);
			$url .= "?" . $get;
		}
		
		curl_setopt($curlproc, CURLOPT_URL, $url);
		
		curl_setopt($curlproc, CURLOPT_ENCODING, ""); // default encoding
		curl_setopt($curlproc, CURLOPT_HTTPHEADER,
			["User-Agent: " . config::USER_AGENT,
			"Accept: */*",
			"Accept-Language: en-US",
			"Accept-Encoding: gzip, deflate, br, zstd",
			"Referer: {$referer}",
			"client-geo-region: global",
			"x-client-version: 8999df28be3f138bf2c646df5d656e4dc6970ba0",
			"DNT: 1",
			"Sec-GPC: 1",
			"Connection: keep-alive",
			"Sec-Fetch-Dest: empty",
			"Sec-Fetch-Mode: cors",
			"Sec-Fetch-Site: same-origin",
			"Priority: u=0",
			"TE: trailers"]
		);
		
		curl_setopt($curlproc, CURLOPT_RETURNTRANSFER, true);
		curl_setopt($curlproc, CURLOPT_SSL_VERIFYHOST, 2);
		curl_setopt($curlproc, CURLOPT_SSL_VERIFYPEER, true);
		curl_setopt($curlproc, CURLOPT_CONNECTTIMEOUT, 30);
		curl_setopt($curlproc, CURLOPT_TIMEOUT, 30);

		$this->backend->assign_proxy($curlproc, $proxy);
		
		$data = curl_exec($curlproc);
		
		if(curl_errno($curlproc)){
			
			throw new Exception(curl_error($curlproc));
		}
		
		curl_close($curlproc);
		return $data;
	}
	
	public function image($get){
		
		if($get["npt"]){
			
			[$filter, $proxy] =
				$this->backend->get(
					$get["npt"],
					"images"
				);
			
			$filter = json_decode($filter, true);
			
		}else{
			
			$search = $get["s"];
			if(strlen($search) === 0){
				
				throw new Exception("Search term is empty!");
			}
			
			$proxy = $this->backend->get_ip();
			
			$filter = [
				"page" => 1,
				"per_page" => 20,
				"query" => $search
			];
			
			// add filters
			if($get["order_by"] != "relevance"){
				
				$filter["order_by"] = $get["order_by"];
			}
			
			if($get["orientation"] != "any"){
				
				$filter["orientation"] = $get["orientation"];
			}
			
			if($get["license"] != "any"){
				
				$filter["plus"] = $get["license"];
			}
		}
		
		$out = [
			"status" => "ok",
			"npt" => null,
			"image" => []
		];
				
		// https://unsplash.com/s/photos/shibuya-night?order_by=latest&orientation=landscape&license=free
		// https://unsplash.com/s/photos/%3Ctest-%3F!-haha == <test ?! haha
		
		// ?page=1&per_page=20&query=japan
		
		try{
			$json =
				$this->get(
					$proxy,
					"https://unsplash.com/napi/search/photos",
					$filter,
					"https://unsplash.com/s/photos/" . rawurlencode(str_replace(" ", "-", $filter["query"])),
				);
			
		}catch(Exception $error){
			
			throw new Exception("Failed to fetch JSON");
		}
		
		$json = json_decode($json, true);
		
		if($json === null){
			
			throw new Exception("Could not decode the JSON payload");
		}
		
		foreach($json["results"] as $image){
			
			$base = explode("?", $image["urls"]["raw"])[0];
			
			if(
				(bool)$image["premium"] ||
				(bool)$image["plus"]
			){
				
				// when we get "premium" images, give
				// 1. highest resolution with watermarks
				// 2. highest resolution without watermarks
				// (if width of image is above 900px, it has watermarks)
				// 3. thumbnail without watermark
				$x900 = $this->imgratio((int)$image["width"], (int)$image["height"], 900);
				$x500 = $this->imgratio((int)$image["width"], (int)$image["height"], 500);
				
				$source = [
					[
						"url" => $base,
						"width" => (int)$image["width"],
						"height" => (int)$image["height"]
					],
					[
						"url" => $base . "?w=900",
						"width" => $x900[0],
						"height" => $x900[1]
					],
					[
						"url" => $base . "?w=500",
						"width" => $x500[0],
						"height" => $x500[1]
					]
				];
			}else{
				
				$x500 = $this->imgratio((int)$image["width"], (int)$image["height"], 500);
				
				// image is free as in freedom(tm)
				$source = [
					[
						"url" => $base,
						"width" => (int)$image["width"],
						"height" => (int)$image["height"]
					],
					[
						"url" => $base . "?w=500",
						"width" => $x500[0],
						"height" => $x500[1]
					]
				];
			}
			
			$title = [];
			
			$image["description"] = trim($image["description"]);
			$image["alt_description"] = trim($image["alt_description"]);
			
			if(!empty($image["description"])){ $title[] = $image["description"]; }
			if(!empty($image["alt_description"])){ $title[] = $image["alt_description"]; }
			
			$title = implode(": ", $title);
			
			$out["image"][] = [
				"title" => $title,
				"source" => $source,
				"url" => "https://unsplash.com/photos/" . $image["slug"]
			];
		}
		
		// next page stuff
		if($filter["page"] < (int)$json["total_pages"]){
			
			$filter["page"]++;
			
			$out["npt"] =
				$this->backend->store(
					json_encode($filter),
					"images",
					$proxy
				);
		}
		
		return $out;
	}
	
	private function imgratio($width, $height, $max_width){
		
		$ratio = $max_width / $width;
		$new_height = floor($height * $ratio);
		
		return [
			$max_width,
			$new_height
		];
	}
}
